cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/concurrency/ExecutionConcurrencyManager.java

```diff
@@ -0,0 +1,271 @@
+    switch (concurrencyType) {
+      case INTER_BROKER_REPLICA:
+        _requestedInterBrokerPartitionMovementConcurrency = concurrency;
+        _interBrokerPartitionMovementConcurrency.replaceAll((k, v) -> concurrency);
+        break;
+      case INTRA_BROKER_REPLICA:
+        _requestedIntraBrokerPartitionMovementConcurrency = concurrency;
+        _intraBrokerPartitionMovementConcurrency.replaceAll((k, v) -> concurrency);
+        break;
+      case LEADERSHIP:
+        _requestedLeadershipMovementConcurrency = concurrency;
+        _leadershipMovementConcurrency.replaceAll((k, v) -> concurrency);
+        break;
+      default:
+        throw new IllegalArgumentException("Unsupported concurrency type " + concurrencyType + " is provided.");
+    }
+    refreshExecutionConcurrencySummary();
+  }
+
+  /**
+   * Set the allowed execution concurrency of a certain concurrency type for a broker
+   * @param brokerId the id of the broker to set allowed concurrency
+   * @param concurrency the allowed concurrency to set
+   * @param concurrencyType the concurrency type of the execution
+   */
+  public synchronized void setExecutionConcurrencyForBroker(int brokerId, Integer concurrency, ConcurrencyType concurrencyType) {
+    sanityCheckRequestedConcurrency(concurrency, concurrencyType);
+    switch (concurrencyType) {
+      case INTER_BROKER_REPLICA:
+        _interBrokerPartitionMovementConcurrency.put(brokerId, concurrency);
+        break;
+      case INTRA_BROKER_REPLICA:
+        _intraBrokerPartitionMovementConcurrency.put(brokerId, concurrency);
+        break;
+      case LEADERSHIP:
+        _leadershipMovementConcurrency.put(brokerId, concurrency);
+        break;
+      default:
+        throw new IllegalArgumentException("Unsupported concurrency type " + concurrencyType + " is provided.");
+    }
+    refreshExecutionConcurrencySummary();
+  }
+
+  /**
+   * Dynamically set the max inter-broker partition movements in cluster
+   * Ensure that the requested max is not greater than the maximum number of allowed movements in cluster.
+   *
+   * @param requestedMaxInterBrokerPartitionMovements The maximum number of concurrent inter-broker partition movements per broker
+   *                                                  (if null, use {@link #_defaultInterBrokerPartitionMovementConcurrency}).
+   */
+  public synchronized void setRequestedMaxInterBrokerPartitionMovements(Integer requestedMaxInterBrokerPartitionMovements) {
+    if (requestedMaxInterBrokerPartitionMovements != null
+        && requestedMaxInterBrokerPartitionMovements > _maxNumClusterMovementConcurrency) {
+      throw new IllegalArgumentException("Attempt to set max inter-broker partition movements [" + requestedMaxInterBrokerPartitionMovements
+                                         + "] to greater than the maximum" + " number of allowed movements in cluster ["
+                                         + _maxNumClusterMovementConcurrency + "].");
+    }
```

Comment:
FWIW, I tend to hate this kind of logic since it ties together two different configs (which in itself is reasonable) whose relationship may not be well-documented in the source of truth (cfg2), and the result of a mistake in the latter is (AFAICT) a completely broken Cruise Control instance.  Is there a reason not to correct the too-large setting (requestedMaxInterBrokerPartitionMovements) to the sanity-check cap (_maxNumClusterMovementConcurrency) and log both this fact and what config name to change if the former was the correct value and the latter the mistake?

Reply:
Your concern totally makes sense, and the way we map cc config to li-cc config can totally cause a pain when missing sufficient description in li-cc...

>  Is there a reason not to correct the too-large setting (requestedMaxInterBrokerPartitionMovements) to the sanity-check cap (_maxNumClusterMovementConcurrency) and log both this fact and what config name to change if the former was the correct value and the latter the mistake?
I think it's because we don't want to silently use a value that not specified by user. E.g. if we just cap the requested value without telling the user, the user may think CC is using their requested value while it actually not. 

I think the best way to do it is to fail the execution, while output the failure reason as part of the command output. That may require another non-trivial refactor though.